apps/examples/smartfs/read_optimization/read_optimization_main.c

```diff
@@ -0,0 +1,273 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name NuttX nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <sys/stat.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <time.h>
+
+#include <tinyara/timer.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define SAMPLE_FILE	"/mnt/sample_file"
+#define TIMER_DEVNAME	"/dev/timer%d"
+
+/****************************************************************************
+ * Private data
+ ****************************************************************************/
+char g_buf[1025];
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: read_test
+ *
+ * Description: Perform a back-to-back read of the entire file's contents,
+ *              multiple times.
+ *
+ ****************************************************************************/
+
+static int read_test(int size)
+{
+	int fd;
+	int ret;
+	int itr = 0;
+
+	while (itr < 100) {
+		fd = open(SAMPLE_FILE, O_RDOK);
+		if (fd < 0) {
+			printf("Unable to open Sample File while testing, Errno = %d\n", errno);
+			return ERROR;
+		}
+
+		do {
+			ret = read(fd, g_buf, size);
```

Comment:
there is no possibility to read partial data from fd?

Reply:
In the scope of this example, a partial read would mean read failure at the FS end,
and for this example, our aim is only simple demonstration.